### PR TITLE
Hive: Introduce HiveMetastoreExtension for Hive tests

### PR DESCRIPTION
**Pull request by nk1506**
CC: @nastra

### CODE REVIEW

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 1]
```diff
@@ -288,7 +288,7 @@ public void createNamespace(Namespace namespace, Map<String, String> meta) {
 
     } catch (AlreadyExistsException e) {
       throw new org.apache.iceberg.exceptions.AlreadyExistsException(
-          e, "Namespace '%s' already exists!", namespace);
+          e, "Namespace already exists: %s", namespace);
```
Comment (nastra) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:291]: unrelated change, please revert this file

File: hive-metastore/src/test/java/org/apache/iceberg/hive/HiveTableBaseTest.java [hunk 2]
```diff
@@ -26,22 +26,34 @@
...
 
   static final String TABLE_NAME = "tbl";
+  static final String DB_NAME = "hivedb";
   static final TableIdentifier TABLE_IDENTIFIER = TableIdentifier.of(DB_NAME, TABLE_NAME);
 
+  @RegisterExtension
+  static final HiveMetastoreExtension HIVE_METASTORE_EXTENSION =
+      new HiveMetastoreExtension(DB_NAME, Collections.emptyMap());
+
+  HiveCatalog catalog;
```
Comment (nastra) [inline hive-metastore/src/test/java/org/apache/iceberg/hive/HiveTableBaseTest.java:55]:
```suggestion
  protected HiveCatalog catalog;
```

File: hive-metastore/src/test/java/org/apache/iceberg/hive/HiveTableBaseTest.java [hunk 3]
```diff
@@ -26,22 +26,34 @@
...
 
-public class HiveTableBaseTest extends HiveMetastoreTest {
+public class HiveTableBaseTest {
 
   static final String TABLE_NAME = "tbl";
+  static final String DB_NAME = "hivedb";
   static final TableIdentifier TABLE_IDENTIFIER = TableIdentifier.of(DB_NAME, TABLE_NAME);
 
+  @RegisterExtension
+  static final HiveMetastoreExtension HIVE_METASTORE_EXTENSION =
```
Comment (nastra) [inline hive-metastore/src/test/java/org/apache/iceberg/hive/HiveTableBaseTest.java:52]:
```suggestion
  protected static final HiveMetastoreExtension HIVE_METASTORE_EXTENSION =
```

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCommitLocks.java [hunk 4]
```diff
@@ -88,13 +95,33 @@ public class TestHiveCommitLocks extends HiveTableBaseTest {
...
+  static final TableIdentifier TABLE_IDENTIFIER = TableIdentifier.of(DB_NAME, TABLE_NAME);
+
+  @RegisterExtension
+  private static final HiveMetastoreExtension HIVE_METASTORE_EXTENSION =
+      new HiveMetastoreExtension(
+          DB_NAME, ImmutableMap.of(HiveConf.ConfVars.HIVE_TXN_TIMEOUT.varname, "1s"));
+
+  private HiveCatalog catalog;
+
+  public void startMetastore() throws Exception {
```
Comment (nastra) [inline hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCommitLocks.java:112]:
```suggestion
  private void startMetastore() throws Exception {
```

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestCachedClientPool.java [hunk 5]
```diff
@@ -28,17 +28,41 @@
...
 import org.apache.iceberg.hive.CachedClientPool.Key;
 import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
 
-public class TestCachedClientPool extends HiveMetastoreTest {
+public class TestCachedClientPool {
+
+  private static final long EVICTION_INTERVAL = TimeUnit.SECONDS.toMillis(10);
```
Comment (nastra) [inline hive-metastore/src/test/java/org/apache/iceberg/hive/TestCachedClientPool.java:44]: can you make this please consistent with other tests? Other tests don't define a constant but directly use 10 millis in the init code

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestCachedClientPool.java [hunk 6]
```diff
@@ -28,17 +28,41 @@
...
+  private HiveConf hiveConf;
+  private static final String DB_NAME = "hivedb";
+
+  @RegisterExtension
+  private static final HiveMetastoreExtension HIVE_METASTORE_EXTENSION =
+      new HiveMetastoreExtension(DB_NAME, Collections.emptyMap());
+
+  @BeforeEach
+  public void initialize() {
+    hiveConf = HIVE_METASTORE_EXTENSION.hiveConf();
```
Comment (nastra) [inline hive-metastore/src/test/java/org/apache/iceberg/hive/TestCachedClientPool.java:54]: why are we initializing the config like that? In all the other places the config is just used directly from the extension: `HIVE_METASTORE_EXTENSION.hiveConf()`

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestCachedClientPool.java [hunk 7]
```diff
@@ -28,17 +28,41 @@
...
 
   @Test
   public void testClientPoolCleaner() throws InterruptedException {
+    CatalogUtil.loadCatalog(
+        HiveCatalog.class.getName(),
+        CatalogUtil.ICEBERG_CATALOG_TYPE_HIVE,
+        ImmutableMap.of(
+            CatalogProperties.CLIENT_POOL_CACHE_EVICTION_INTERVAL_MS,
+            String.valueOf(EVICTION_INTERVAL)),
+        hiveConf);
```
Comment (nastra) [inline hive-metastore/src/test/java/org/apache/iceberg/hive/TestCachedClientPool.java:65]:
```suggestion
        HIVE_METASTORE_EXTENSION.hiveConf());
```

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestCachedClientPool.java [hunk 8]
```diff
@@ -28,17 +28,41 @@
...
+  }
 
   @Test
   public void testClientPoolCleaner() throws InterruptedException {
+    CatalogUtil.loadCatalog(
+        HiveCatalog.class.getName(),
+        CatalogUtil.ICEBERG_CATALOG_TYPE_HIVE,
+        ImmutableMap.of(
+            CatalogProperties.CLIENT_POOL_CACHE_EVICTION_INTERVAL_MS,
+            String.valueOf(EVICTION_INTERVAL)),
```
Comment (nastra) [inline hive-metastore/src/test/java/org/apache/iceberg/hive/TestCachedClientPool.java:64]:
```suggestion
            String.valueOf(TimeUnit.SECONDS.toMillis(10))),
```

**Review by nastra — CHANGES_REQUESTED** [review]

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestCachedClientPool.java [hunk 9]
```diff
@@ -28,30 +28,49 @@
...
+  private static final String DB_NAME = "hivedb";
+
+  @RegisterExtension
+  private static final HiveMetastoreExtension HIVE_METASTORE_EXTENSION =
+      new HiveMetastoreExtension(DB_NAME, Collections.emptyMap());
 
   @Test
   public void testClientPoolCleaner() throws InterruptedException {
-    CachedClientPool clientPool = new CachedClientPool(hiveConf, Collections.emptyMap());
+    CatalogUtil.loadCatalog(
```
Comment (nastra) [inline hive-metastore/src/test/java/org/apache/iceberg/hive/TestCachedClientPool.java:51]: do we need to load the catalog here if it's not being used anywhere?

Comment (nastra) [reply hive-metastore/src/test/java/org/apache/iceberg/hive/TestCachedClientPool.java:51]: you should be able to pass the eviction interval like this (without having to initialize the catalog:

```
CachedClientPool clientPool =
        new CachedClientPool(
            hiveConf,
            ImmutableMap.of(
                CatalogProperties.CLIENT_POOL_CACHE_EVICTION_INTERVAL_MS,
                String.valueOf(EVICTION_INTERVAL)));
```

**Review by nastra — APPROVED** [review]
thanks @nk1506, this LGTM

**Comment by ajantha-bhat** [comment]
Let me take a look

File: hive-metastore/src/test/java/org/apache/iceberg/hive/HiveMetastoreExtension.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+import java.util.Map;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.metastore.HiveMetaStoreClient;
+import org.apache.hadoop.hive.metastore.api.Database;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.junit.jupiter.api.extension.AfterEachCallback;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+
+public class HiveMetastoreExtension implements BeforeEachCallback, AfterEachCallback {
```
Comment (ajantha-bhat) [inline hive-metastore/src/test/java/org/apache/iceberg/hive/HiveMetastoreExtension.java:30]: Should this be `BeforeAllCallback` and `AfterEachCallback` ?

I don't think starting and stopping metastore for each test case is a good idea

File: hive-metastore/src/test/java/org/apache/iceberg/hive/HiveCreateReplaceTableTest.java [hunk 11]
```diff
@@ -60,8 +66,23 @@ public class HiveCreateReplaceTableTest extends HiveMetastoreTest {
...
 
+  @RegisterExtension
+  private static final HiveMetastoreExtension HIVE_METASTORE_EXTENSION =
+      new HiveMetastoreExtension(DB_NAME, Collections.emptyMap());
+
+  private HiveCatalog catalog;
+
   @BeforeEach
   public void createTableLocation() throws IOException {
+    catalog =
```
Comment (ajantha-bhat) [inline hive-metastore/src/test/java/org/apache/iceberg/hive/HiveCreateReplaceTableTest.java:77]: I think we should initialize the catalog only once. Not for each tests as nothing changes between tests related to catalog config.

File: hive-metastore/src/test/java/org/apache/iceberg/hive/HiveMetastoreExtension.java [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    HiveConf hiveConfWithOverrides = new HiveConf(TestHiveMetastore.class);
+    if (hiveConfOverride != null) {
+      for (Map.Entry<String, String> kv : hiveConfOverride.entrySet()) {
+        hiveConfWithOverrides.set(kv.getKey(), kv.getValue());
+      }
+    }
+
+    metastore.start(hiveConfWithOverrides);
+    metastoreClient = new HiveMetaStoreClient(hiveConfWithOverrides);
+
```
Comment (ajantha-bhat) [inline hive-metastore/src/test/java/org/apache/iceberg/hive/HiveMetastoreExtension.java:53]: nit: we can remove the new line

File: hive-metastore/src/test/java/org/apache/iceberg/hive/HiveCreateReplaceTableTest.java (hunk 11 above)
Comment (ajantha-bhat) [reply hive-metastore/src/test/java/org/apache/iceberg/hive/HiveCreateReplaceTableTest.java:77]: Also can we move this to `HiveMetastoreExtension` and get the catalog from there? It can help in duplicating this code in each test class which uses this extension.

File: hive-metastore/src/test/java/org/apache/iceberg/hive/HiveMetastoreExtension.java (hunk 10 above)
Comment (nk1506) [reply hive-metastore/src/test/java/org/apache/iceberg/hive/HiveMetastoreExtension.java:30]: @ajantha-bhat , we are planning to extend [CatalogTests](https://github.com/apache/iceberg/blob/main/core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java) for Hive too. It expects all the catalog related infos to be dropped after each tests.

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestCachedClientPool.java [hunk 13]
```diff
@@ -28,30 +28,46 @@
...
+                    CachedClientPool.extractKey(null, HIVE_METASTORE_EXTENSION.hiveConf())));
     TimeUnit.MILLISECONDS.sleep(EVICTION_INTERVAL - TimeUnit.SECONDS.toMillis(2));
     HiveClientPool clientPool2 = clientPool.clientPool();
     assertThat(clientPool2).isSameAs(clientPool1);
     TimeUnit.MILLISECONDS.sleep(EVICTION_INTERVAL + TimeUnit.SECONDS.toMillis(5));
     assertThat(
             CachedClientPool.clientPoolCache()
-                .getIfPresent(CachedClientPool.extractKey(null, hiveConf)))
+                .getIfPresent(
+                    CachedClientPool.extractKey(null, HIVE_METASTORE_EXTENSION.hiveConf())))
```
Comment (ajantha-bhat) [inline hive-metastore/src/test/java/org/apache/iceberg/hive/TestCachedClientPool.java:70]: nit: some test we are extracting to a variable and in some test we are using directly. Let us have a unified style.

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCommitLocks.java [hunk 14]
```diff
@@ -88,13 +95,33 @@ public class TestHiveCommitLocks extends HiveTableBaseTest {
...
+  static final TableIdentifier TABLE_IDENTIFIER = TableIdentifier.of(DB_NAME, TABLE_NAME);
+
+  @RegisterExtension
+  private static final HiveMetastoreExtension HIVE_METASTORE_EXTENSION =
+      new HiveMetastoreExtension(
+          DB_NAME, ImmutableMap.of(HiveConf.ConfVars.HIVE_TXN_TIMEOUT.varname, "1s"));
+
+  private HiveCatalog catalog;
+
+  private void startMetastore() throws Exception {
```
Comment (ajantha-bhat) [inline hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCommitLocks.java:112]: this method is not starting the `metastore`

File: hive-metastore/src/test/java/org/apache/iceberg/hive/HiveCreateReplaceTableTest.java (hunk 11 above)
Comment (nastra) [reply hive-metastore/src/test/java/org/apache/iceberg/hive/HiveCreateReplaceTableTest.java:77]: > Also can we move this to `HiveMetastoreExtension` and get the catalog from there? It can help in duplicating this code in each test class which uses this extension.

I don't think the catalog initialization should be moved into `HiveMetastoreExtension`. The sole purpose of `HiveMetastoreExtension` is to provide the metastore part

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCommitLocks.java (hunk 14 above)
Comment (nastra) [reply hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCommitLocks.java:112]: indeed I missed this, let's rename it to ìnitCatalog()`

File: hive-metastore/src/test/java/org/apache/iceberg/hive/HiveCreateReplaceTableTest.java (hunk 11 above)
Comment (ajantha-bhat) [reply hive-metastore/src/test/java/org/apache/iceberg/hive/HiveCreateReplaceTableTest.java:77]: @nastra: Thoughts?

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestCachedClientPool.java (hunk 13 above)
Comment (nk1506) [reply hive-metastore/src/test/java/org/apache/iceberg/hive/TestCachedClientPool.java:70]: At other test `hiveConf` was being used at multiple places, that's the reason I created a local variable.

File: hive-metastore/src/test/java/org/apache/iceberg/hive/HiveCreateReplaceTableTest.java (hunk 11 above)
Comment (nastra) [reply hive-metastore/src/test/java/org/apache/iceberg/hive/HiveCreateReplaceTableTest.java:77]: I've replied already above

Comment (ajantha-bhat) [reply hive-metastore/src/test/java/org/apache/iceberg/hive/HiveCreateReplaceTableTest.java:77]: Sorry, I wanted your thoughts on this comment
https://github.com/apache/iceberg/pull/9282#discussion_r1423803359

File: hive-metastore/src/test/java/org/apache/iceberg/hive/HiveMetastoreExtension.java (hunk 10 above)
Comment (ajantha-bhat) [reply hive-metastore/src/test/java/org/apache/iceberg/hive/HiveMetastoreExtension.java:30]: If we drop all namespaces and tables after each test, catalog tests can work fine IMO. 

I think starting and stopping metastore for each tests is still expensive. 

If others (@nastra) is ok with this change, Then fine.

Comment (nastra) [reply hive-metastore/src/test/java/org/apache/iceberg/hive/HiveMetastoreExtension.java]: I wrongly remembered that we needed this for https://github.com/apache/iceberg/pull/8918, but in fact we can just init the catalog before each method in `TestHiveCatalog` for https://github.com/apache/iceberg/pull/8918, which means that we can indeed start/stop the metastore using the BeforeAllCallback/AfterAllCallback.

I have pushed a commit that addresses this

**Review by nastra — APPROVED** [review]

**Review by ajantha-bhat — APPROVED** [review]
Now it is LGTM.

File: hive-metastore/src/test/java/org/apache/iceberg/hive/HiveMetastoreExtension.java (hunk 10 above)
Comment (nk1506) [reply hive-metastore/src/test/java/org/apache/iceberg/hive/HiveMetastoreExtension.java]: Unless I am missing something using `BeforeAllCallback, AfterAllCallback`  with HiveMetastoreExtension will stop us using the `CatalogTests`. I mean it may ask many custom changes with `TestHiveCatalog`. 
Attaching a reference for the same.

Comment (ajantha-bhat) [reply hive-metastore/src/test/java/org/apache/iceberg/hive/HiveMetastoreExtension.java]: @nk1506: 
we can add a proper teardown in the beforeEach() like Nessie does when we extend it.
https://github.com/apache/iceberg/blob/0c5b87a27e0564ddaae2fa1cb93c32dd7a5e8681/nessie/src/test/java/org/apache/iceberg/nessie/TestNessieCatalog.java#L97

I think Nessie server is up only once. Instead of per testcase.
https://github.com/projectnessie/nessie/blob/main/servers/jax-rs-testextension/src/main/java/org/projectnessie/jaxrs/ext/NessieJaxRsExtension.java

Comment (nk1506) [reply hive-metastore/src/test/java/org/apache/iceberg/hive/HiveMetastoreExtension.java]: With Hive it can't drop namespace unless all the tables have been dropped. It throws error like.
`InvalidOperationException(message:Database newdb is not empty. One or more tables exist.)`
I had to [revert](https://github.com/apache/iceberg/pull/8918/files#diff-17a52cf779b5dc58f09db91962cb69514957356c8fe74dc8df36d9dbec8074d7R30) and added it as teardown of metastore after every tests.
